rfd/0159-password-status.md

```diff
@@ -0,0 +1,125 @@
+OldAndNewPass --> ChangePass
+ChangePass --> End
+MFAAvail --> |No| MFAReq{MFA required?}
+MFAReq --> |No| OldAndNewPass
+MFAReq --> |Yes| ContactAdmin[Please contact the admin]
+ContactAdmin --> End([End])
+```
+
+For clarity, error cases (such as password mismatch or inability to confirm user identity) were omitted from the above diagram.
+
+### Recognizing Users Who Configured Their Passwords
+
+Ultimately, we would like to have information about password status for as many users as possible, for the following reasons:
+
+1. To explicitly tell the user that they have a password configured — for example, on the account settings page.
+2. To support any other future feature that might require knowing this information.
+
+To do this, we will extend the [`UserSpecV2`](https://github.com/gravitational/teleport/blob/6103a2b1eff2a6bece80ab138ebe68b4ba041a91/api/proto/teleport/legacy/types/types.proto#L3185) message by adding a `PasswordState` field:
+
+```proto
+message UserSpecV2 {
+  // ...existing fields...
+
+  PasswordState PasswordState = 11 [(gogoproto.jsontag) = "password_state,omitempty"];
+}
+
+enum PasswordState {
+  // Unable to tell whether the password has been configured.
+  PASSWORD_STATE_UNSPECIFIED = 0;
+  // Password is known to be not configured.
+  PASSWORD_STATE_UNSET = 1;
+  // Password is known to be configured.
+  PASSWORD_STATE_SET = 2;
+}
+```
+
+The flag will be stored under the `/web/users/<username>/params` key, along with the rest of `UserSpecV2` fields.
+
+`PASSWORD_STATE_UNSPECIFIED` is deliberately set to 0, since this is the default value when there's no information about the password state in the database.
+
+The state changes to `PASSWORD_STATE_UNSET` whenever another user (or cluster onboarding flow) creates a password reset token for a given user. These are scenarios when the system becomes confident that the user has no password. As a consequence, users created after this change is made will never have their password in an unspecified state.
+
+The state changes to `PASSWORD_STATE_SET` whenever user sets/resets their password or successfully signs in using a password. The last scenario, in particular, will allow us to gradually fill in information about existing users that have passwords.
+
+> [!WARNING]
+> The PasswordState flag should _not_ be used for authentication itself. Attempting to shortcut password verification may open a timing attack vulnerability.
+
+### Returning Password State to the Account Settings UI
+
+The old account settings page had separate paths for managing passwords and MFA devices, so there is no endpoint that would return information about all authentication methods specifically for this page. There is only an endpoint that returns a list of MFA devices.
+
+Since we already return `authType` in `/v1/webapi/sites/<site>/context`, it makes sense to extend the returned data structure with the `PasswordStatus` flag, thus making it appear in the `UserContext` structure on the frontend.
```

Comment:
Is there an RPC that backs this call? Are we doing proto API changes as a result of this?